File: mahaam-api-cs/Src/Infra/Exceptions.cs
using System.Net;

namespace Mahaam.Infra;

public abstract class AppException : Exception
{
	public string? Key { get; }
	public int HttpCode { set; get; }
	public AppException(int httpCode) : base() { HttpCode = httpCode; }
	public AppException(string message, int httpCode, string? key = null) : base(message) { Key = key; HttpCode = httpCode; }
}

public class UnauthorizedException(string message) : AppException(message, (int)HttpStatusCode.Unauthorized)
{
}

public class ForbiddenException(string message) : AppException(message, (int)HttpStatusCode.Forbidden)
{
}

public class LogicException(string message, string? key = null) : AppException(message, (int)HttpStatusCode.Conflict, key)
{
}

public class NotFoundException(string message) : AppException(message, (int)HttpStatusCode.NotFound)
{
}

public class InputException(string message) : AppException(message, (int)HttpStatusCode.BadRequest)
{
}

File: mahaam-api-py/infra/exceptions.py
from infra import http


class AppException(Exception):
    def __init__(self, message=None, http_code=None, key=None):
        super().__init__(message)
        self.key = key
        self.http_code = http_code


class UnauthorizedException(AppException):
    def __init__(self, message):
        super().__init__(message, http.UNAUTHORIZED)


class ForbiddenException(AppException):
    def __init__(self, message):
        super().__init__(message, http.FORBIDDEN)


class LogicException(AppException):
    def __init__(self, message, key=None):
        super().__init__(message, http.CONFLICT, key)


class NotFoundException(AppException):
    def __init__(self, message):
        super().__init__(message, http.NOT_FOUND)


class InputException(AppException):
    def __init__(self, message):
        super().__init__(message, http.BAD_REQUEST)
